### PR TITLE
[JENKINS-69339] Add support for badge icons in Management links

### PR DESCRIPTION
**Pull request by mawinter69**
Allow to define a badge icon and a tooltip for all management links.

Adjust PluginsLink to the new method
Add implementations for OldDataMonitor and NodesLink

See [JENKINS-69339](https://issues.jenkins.io/browse/JENKINS-69339).

### Proposed changelog entries

* Add support for badge icons in Management links

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6995"></a>

### CODE REVIEW

File: core/src/main/java/hudson/model/ManagementLink.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/ManagementLink.java:225]: This should probably be limited to returning numbers, or is there a real use case for non-numeric badge values that wouldn't quickly overwhelm the badge size?

File: core/src/main/java/hudson/model/ManagementLink.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/ManagementLink.java:235]: Seems weird to require an empty string when there's no content. Why?

File: core/src/main/java/hudson/model/ManagementLink.java
Comment (mawinter69) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: I was thinking for my agent maintenance plugin to display something like 5/12 with a tooltip "5 active maintenance windows 12 total"

Comment (mawinter69) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: Another use case might be to show a "!" to indicate that some action needs to be taken

Comment (daniel-beck) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: Interesting, good point with 5/12.

Still, I wonder whether we could define this API so that it prevents weird abuse (like actual text, as happened in `groovy-postbuild` with build badges)

Comment (mawinter69) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: when returning 1000000 it will probably also look weird. One could also think of an array if integers and take only the first 2 or 3 elements of the array and separate it with "/".
Or just take a substring of the first 8 to 10 characters.

Comment (mawinter69) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: Some samples how it would look:

Long string without blanks
image

String with blanks leads to a line break
image

image

image

Comment (daniel-beck) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: > when returning 1000000 it will probably also look weird.

Realistically, no number will grow that large. The upper bounds should be hundreds of plugins need updating, (tens of) thousands of jobs fail to load, so that's ~5 digits?

Comment (daniel-beck) [reply core/src/main/java/hudson/model/ManagementLink.java:225]: Also, if it's a string you'll need to define whether it supports HTML or not. Right now it would (and will need to escape any user content).

**Comment by daniel-beck** [comment]
The weird thing about the existing badge is that it only exists for a feature that does not have a corresponding admin monitor (plugin updates). My guess is that it's probably deliberately so, because there are always plugin updates, especially since JEP-229, so having an admin monitor that's always on would be spammy and quickly ignored by many admins.

Other features like Old Data Monitor do have an admin monitor when they have useful content. Usually because this is something that _needs_ attention (in that case, loss of configuration data, loss of functionality, etc.).

If this is added, we need to provide guidelines to explain when you would have the admin monitor, when you would have the badge, and when you would have both.

An interesting case is In-Process Script Approval (`script-security` plugin): It currently has a badge-like feature (the label changes). This feature should be migrated to badges if this is merged (would make a nice downstream PR for this change), and I wonder whether an admin monitor would make sense there as well. Could probably go either way.

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:89]: This is now (once re-merged with `master`) a non-HTML tooltip. That seems fine.

**Review by daniel-beck — COMMENTED** [review]
I really like this PR. It cleans up the existing tooltip for plugin updates and adds useful additional tooltips (although I'm not entirely sure the node one is useful as-is, given agents on a schedule etc.).

An open question remains about the guidelines for plugin developers on how to use this feature. We should have documentation when to (not) use this ready at the same time we publish this API addition. If we keep the `String` type, the docs should also explain it's for short labels only (probably mostly numeric; e.g., `4 of 12` might be OK, `21 available updates` is not). This should be added to Javadoc and/or https://plugins.jenkins.io/design-library/

Otherwise (and assuming a quick re-test confirms the non-HTML tooltip behavior) this looks good 👍

**Comment by daniel-beck** [comment]
> assuming a quick re-test confirms the non-HTML tooltip behavior

Confirmed, non-HTML tooltip and content when merged with `master`.

**Comment by mawinter69** [comment]
I'm wondering if there are other places where such badges might be useful. If so it might make sense to create a Badge class that contains the data, tooltip and maybe a severity so the badge can have different colors. Even when not used somewhere else it might help to avoid running some calculations twice (once for the badge and once for the tooltip.
For severity, e.g. when there are plugin updates that fix security issues display it in red, otherwise in orange.

**Comment by mawinter69** [comment]
> (although I'm not entirely sure the node one is useful as-is, given agents on a schedule etc.).

The agents on a schedule wouldn't be a problem, but the ondemand would show up whenever they have nothing to do and thus disconnect. Maybe restrict the check to agents that use the always online strategy which is the default and most people use I guess.

**Comment by daniel-beck** [comment]
> avoid running some calculations twice (once for the badge and once for the tooltip

For expensive calculations, could be handled by individual `ManagementLink` in a cache. I don't think there are _that_ many cases where computation would be expensive?

Feel free to sketch this out, might be reasonable for future extension (e.g. severity/color), but IMO all that's missing here is related documentation.

> Maybe restrict the check to agents that use the always online strategy which is the default and most people use I guess.

Right, something along those lines. Might be simpler to remove this from the PR for now and look into it as followup. Migrating Manage Plugins to this is already pretty nice cleanup, and Manage Old Data is a reasonable new use.

**Comment by mawinter69** [comment]
I removed the badge for the nodes now, there is no way of reliably determining this information, e.g. when the agent-maintenance enabled for an agent, which is proxying the retentionstrategy 
Added a Badge class to wrap all the information
Plugin update badge is now red when updates fix security vulnerabilities and orange otherwise. Tooltip shows the actual numbers for total updates and updates with security fixes. One could think of extending the badge/tooltip also for plugins that have incompatible changes

File: core/src/main/java/hudson/model/ManagementLink.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/ManagementLink.java:224]: One drawback of the new `Badge` type is that it requires a new core dependency. Previously, the `String` based API could be implemented even by plugins with an old core dependency.

I don't think this change must be reverted, but considering simple page navigation shouldn't be expensive anyway, the argument for one call instead of two isn't _that_ strong (expensive computation should be done elsewhere and cached values shown, kind of like available updates).

File: core/src/main/java/jenkins/management/PluginsLink.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/PluginsLink.java:96]: Interesting approach, I kinda like it. Less red is better when there's no immediate need to act, especially with a badge that is practically always visible.

FWIW I look forward to reading in docs about the interaction between admin monitors and badges here :)

File: core/src/main/java/jenkins/management/Badge.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/management/Badge.java:60]: Convention is for these to be ALLCAPS.

(I expected there to be a static check for this, kinda surprised.)

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (daniel-beck) [inline core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:90]: Does this handle `null` tooltips correctly? (i.e., no tooltip) There's no `@NonNull` annotation so I expect them to be optional.

(Untested, planning to check later unless someone else is faster.)

Comment (daniel-beck) [inline core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:90]: No `@NonNull` annotation on `severity`, and no fallback handling for `null` parameter, but I don't think the result looks reasonable when there's none.

(Untested, planning to check later unless someone else is faster.)

File: war/src/main/less/modules/section.less
Comment (daniel-beck) [inline war/src/main/less/modules/section.less:113]: Does this only affect Manage Jenkins badges? `.jenkins-section` looks very generic.

**Review by daniel-beck — DISMISSED** [review]

File: core/src/main/java/jenkins/management/Badge.java
Comment (timja) [inline core/src/main/java/jenkins/management/Badge.java:61]: what do you mean by blanks?

File: core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly
Comment (timja) [reply core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:94]: The constructor for badge has nonnull on tooltip now

Comment (timja) [reply core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:94]: severity is now non null

**Review by timja — APPROVED** [review]
LGTM just one wording question in the Javadoc but no blockers, tested locally.

File: core/src/main/java/jenkins/management/Badge.java
Comment (mawinter69) [reply core/src/main/java/jenkins/management/Badge.java:61]: A whitespace can lead to a line break getting rendered. See the first outdated conversation where I added some screenshots

Comment (timja) [reply core/src/main/java/jenkins/management/Badge.java:61]: what about?

```suggestion
     *      Keep it short, ideally just a number. More than 6 or 7 characters do not look good. Avoid spaces as they will lead to line breaks.
```

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
> [timja](https://github.com/timja) dismissed [daniel-beck](https://github.com/daniel-beck)’s [stale review](https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1223585941) Dec 31, 2022
> addressed

In multiple comments I kept requesting developer documentation. Could you point me to it?

https://github.com/jenkinsci/jenkins/pull/6995#issuecomment-1216306653:

> If this is added, we need to provide guidelines to explain when you would have the admin monitor, when you would have the badge, and when you would have both.

https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1219211363:

> An open question remains about the guidelines for plugin developers on how to use this feature. We should have documentation when to (not) use this ready at the same time we publish this API addition. If we keep the `String` type, the docs should also explain it's for short labels only (probably mostly numeric; e.g., `4 of 12` might be OK, `21 available updates` is not). This should be added to Javadoc and/or https://plugins.jenkins.io/design-library/

https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1223585941:

> FWIW I look forward to reading in docs about the interaction between admin monitors and badges here :)

**Comment by timja** [comment]
> In multiple comments I kept requesting developer documentation. Could you point me to it?

https://github.com/jenkinsci/jenkins/blob/a6d542509799d1b70610406c7750f2440c44e8b2/core/src/main/java/jenkins/management/Badge.java#L33-L50

https://github.com/jenkinsci/jenkins/blob/a6d542509799d1b70610406c7750f2440c44e8b2/core/src/main/java/hudson/model/ManagementLink.java#L218-L223

**Comment by daniel-beck** [comment]
Thanks, missed it (or forgot in the mean time) 👍

**Pull request merged**
